javalib/src/main/scala/java/util/RedBlackTree.scala

```diff
@@ -0,0 +1,1019 @@
+   * tree object keeps track of its size. Therefore, while obtaining the size
+   * of the whole tree is O(1), knowing the number of entries inside a range is
+   * O(n) on the size of the range.
+   */
+
+  final class Tree[A, B](var root: Node[A, B], var size: Int) {
+    def treeCopy(): Tree[A, B] = new Tree(copyTree(root), size)
+  }
+
+  object Tree {
+    def empty[A, B]: Tree[A, B] = new Tree(null, 0)
+  }
+
+  final class Node[A, B](
+      val key: A,
+      private[RedBlackTree] var value: B,
+      private[RedBlackTree] var red: Boolean,
+      private[RedBlackTree] var left: Node[A, B],
+      private[RedBlackTree] var right: Node[A, B],
+      private[RedBlackTree] var parent: Node[A, B]
+  ) extends Map.Entry[A, B] {
+    def getKey(): A = key
+
+    def getValue(): B = value
+
+    def setValue(v: B): B = {
+      val oldValue = value
+      value = v
+      oldValue
+    }
+
+    override def equals(that: Any): Boolean = that match {
+      case that: Map.Entry[_, _] =>
+        Objects.equals(getKey(), that.getKey()) &&
+        Objects.equals(getValue(), that.getValue())
+      case _ =>
+        false
+    }
+
+    override def hashCode(): Int =
+      Objects.hashCode(key) ^ Objects.hashCode(value)
+
+    override def toString(): String =
+      "" + getKey + "=" + getValue
+
+    @inline private[RedBlackTree] def isRoot: Boolean =
+      parent eq null
+
+    @inline private[RedBlackTree] def isLeftChild: Boolean =
+      (parent ne null) && (parent.left eq this)
+
+    @inline private[RedBlackTree] def isRightChild: Boolean =
+      (parent ne null) && (parent.right eq this)
+  }
+
+  object Node {
+    @inline
+    def apply[A, B](key: A, value: B, red: Boolean, left: Node[A, B],
+        right: Node[A, B], parent: Node[A, B]): Node[A, B] = {
+      new Node(key, value, red, left, right, parent)
+    }
+
+    @inline
+    def leaf[A, B](key: A, value: B, red: Boolean, parent: Node[A, B]): Node[A, B] =
+      new Node(key, value, red, null, null, parent)
+  }
+
+  // ---- comparator helper ----
+
+  @inline
+  private[this] def compare[A](key1: Any, key2: A)(
+      implicit comp: Comparator[_ >: A]): Int = {
+    /* The implementation of `compare` and/or its generic bridge may perform
+     * monomorphic casts that can fail. This is according to spec for TreeSet
+     * and TreeMap.
+     */
+    comp.asInstanceOf[Comparator[Any]].compare(key1, key2)
+  }
+
+  // ---- getters ----
+
+  private def isRed(node: Node[_, _]): Boolean = (node ne null) && node.red
+
+  private def isBlack(node: Node[_, _]): Boolean = (node eq null) || !node.red
+
+  // A red node is non-null and not the root, so node.parent exists and is not null
+  private def redIsLeft(node: Node[_, _]): Boolean = node eq node.parent.left
+
+  // A red node is non-null and not the root, so node.parent exists and is not null
+  private def redIsRight(node: Node[_, _]): Boolean = node eq node.parent.right
```

Comment:
These two helpers should go away now and be replaced by `node.isLeftChild` and `node.isRightChild` respectively.

Reply:
Done.